node/network/availability-distribution/src/lib.rs

```diff
@@ -0,0 +1,1552 @@
+				}
+				FromOverseer::Signal(OverseerSignal::ActiveLeaves(ActiveLeavesUpdate {
+					activated,
+					deactivated,
+				})) => {
+					for relay_parent in activated {
+						trace!(target: TARGET, "Start {:?}", relay_parent);
+						let per_relay_parent = state
+							.per_relay_parent
+							.entry(relay_parent.clone())
+							.or_default();
+						let validators = query_validators(&mut ctx, relay_parent).await?;
+						per_relay_parent.validator_index = obtain_our_validator_index::<Context>(
+							&validators,
+							self.keystore.clone(),
+						);
+						per_relay_parent.validators = validators;
+					}
+					for relay_parent in deactivated {
+						trace!(target: TARGET, "Stop {:?}", relay_parent);
+					}
+				}
+				FromOverseer::Signal(OverseerSignal::BlockFinalized(_)) => {}
+				FromOverseer::Signal(OverseerSignal::Conclude) => {
+					trace!(target: TARGET, "Conclude");
+					return Ok(());
+				}
+			}
+		}
+	}
+}
+
+impl<Context> Subsystem<Context> for AvailabilityDistributionSubsystem
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage> + Sync + Send,
+{
+	fn start(self, ctx: Context) -> SpawnedSubsystem {
+		SpawnedSubsystem {
+			name: "availability-distribution",
+			future: Box::pin(async move { self.run(ctx) }.map(|_| ())),
+		}
+	}
+}
+
+/// Obtain all live candidates based on an iterator of relay heads.
+async fn query_live_candidates_without_ancestors<Context>(
+	ctx: &mut Context,
+	relay_parents: impl IntoIterator<Item = H256>,
+) -> Result<HashSet<CommittedCandidateReceipt>>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	let iter = relay_parents.into_iter();
+	let hint = iter.size_hint();
+
+	let mut live_candidates = HashSet::with_capacity(hint.1.unwrap_or(hint.0));
+	for relay_parent in iter {
+		let paras = query_para_ids(ctx, relay_parent.clone()).await?;
+		for para in paras {
+			if let Some(ccr) = query_pending_availability(ctx, relay_parent, para).await? {
+				live_candidates.insert(ccr);
+			}
+		}
+	}
+	Ok(live_candidates)
+}
+
+/// Obtain all live canddidates based on an iterator or relay heads including `k` ancestors.
+///
+/// Relay parent.
+async fn query_live_candidates<Context>(
+	ctx: &mut Context,
+	state: &mut ProtocolState,
+	relay_parents: impl IntoIterator<Item = H256>,
+) -> Result<HashMap<H256, CommittedCandidateReceipt>>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	let iter = relay_parents.into_iter();
+	let hint = iter.size_hint();
+
+	let capacity = hint.1.unwrap_or(hint.0) * (1 + AvailabilityDistributionSubsystem::K);
+	let mut live_candidates = HashMap::<H256, CommittedCandidateReceipt>::with_capacity(capacity);
+
+	for relay_parent in iter {
+		// direct candidates for one relay parent
+		let receipts =
+			query_live_candidates_without_ancestors(ctx, iter::once(relay_parent.clone())).await?;
+		live_candidates.extend(iter::once(relay_parent.clone()).zip(receipts.into_iter()));
```

Comment:
is this a correct use of `zip`?  This might take only the first entry from `receipts`, which is a `HashSet`